[util,doc] Add codeblock support to CMDGEN

hcallahan-lowrisc opened the pull request:
Since CMDGEN currently uses HTML-comments, if you wrap these comments inside a markdown codeblock to gain formatting for the result of your command, the HTML comments will also be rendered in the codeblock.

This PR adds an optional argument mode to CMDGEN to insert the codeblocks for you, only this time inside the HTML braces.
Currently, if you malform the optional arguments, the errors aren't very nice. This could be improved with this implementation.

As a demonstrator, replace the hardcoded content of the ipgen.py tool in its README with the CMDGEN-generated content. Use the 'console' formatter to clearly show that we are printing CLI output here.

hcallahan-lowrisc committed:
[util] Add support for codeblock argument to CMDGEN tool

Markdown will render the HTML comment if it is inside a codeblock, so
this is one way around that issue.

Signed-off-by: Harry Callahan <hcallahan@lowrisc.org>

hcallahan-lowrisc committed:
[doc] Switch ipgen README to use CMDGEN for CLI help strings

Signed-off-by: Harry Callahan <hcallahan@lowrisc.org>

hcallahan-lowrisc committed:
[doc] Run CMDGEN **/* -> Remove stray newline at end of output

The previous change to the cmdgen tool cleaned this up.

Signed-off-by: Harry Callahan <hcallahan@lowrisc.org>

HU90m commented on util/cmdgen.py:
```diff
@@ -89,7 +123,7 @@ def cmdgen_rewrite_md(filepath: Path, dry_run: bool, update: bool) -> bool:
 
         # Run the found-command in a subshell
-        res = subprocess.run(cmd, shell=True, capture_output=True)
+        res = subprocess.run(" ".join(cmd_args.cmd), shell=True, capture_output=True)
```
```suggestion
        res = subprocess.run(" ".join(cmd_args.cmd), shell=True, capture_output=True , env={"COLUMNS": "80"})
```

Makes slightly more deterministic

hcallahan-lowrisc committed:
[util] Fix CMDGEN subprocess to use COLUMNS=80

CLI output can be dependent on the width of the terminal it outputs into. Fix this to be more deterministic

Co-authored-by: Hugo McNally <45573837+HU90m@users.noreply.github.com>
Signed-off-by: Harry Callahan <hcallahan@lowrisc.org>

hcallahan-lowrisc committed:
[reggen] Add newline to end of no-regs message

Signed-off-by: Harry Callahan <hcallahan@lowrisc.org>

hcallahan-lowrisc commented:
As part of this PR, I noticed that the output from python's argparse changes at some point between 3.8 and 3.12, with the string "optional arguments" changing to "options". This is a non-functional change, but if you run `./util/cmdgen.py --update` with a newer version of python than is used in CI, it will change the content, and if you commit it will then fail the CI check.

hcallahan-lowrisc commented:
@jwnrt

nbdd0121 commented:
Since CMDGEN runs shell, can we do things more akin to Unix tool philosophy and create a script (`codewrap.py` or `codewrap.sh`) and then pipe the output to it? E.g.

```markdown
<!-- BEGIN CMDGEN util/ipgen.py generate --help | util/codewrap.py --type console -->
```

hcallahan-lowrisc commented:
> 

Yeah I like this more. Thanks for the idea. We should definitely document the cmdgen tools with some examples of solving common problems (such as this), so a contributor can just copy-paste a solution without too much thought.

hcallahan-lowrisc commented:
We should also consider if stdout/stderr handling if we are using pipes in a CMDGEN invocation

nbdd0121 commented:
Or we could just do
```markdown
<!-- BEGIN CMDGEN echo '```console`' && util/ipgen.py generate --help && echo '```' -->
```

hcallahan-lowrisc commented:
> Or we could just do
> 
> ```
> <!-- BEGIN CMDGEN echo '```console' && util/ipgen.py generate --help && echo '```' -->
> ```

IMO this is ugly, but will hear arguments

jwnrt commented:
> IMO this is ugly, but will hear arguments

My view is this isn't very pretty, but I don't think it's worth adding the shorthand flag to cmdgen to avoid

rswarbrick commented:
Hi @hcallahan-lowrisc! Is this PR something that should be resurrected? If not, should it be closed?

hcallahan-lowrisc commented:
Closing as it's not really that important, but could be brought back in the future as a cleanup.
